### PR TITLE
Fix retrieving blocks in 3rd front camera view.

### PR DESCRIPTION
**Pull request by lhofhansl**
Not sure it's even worth the effort, but I noticed that in 3rd front camera view the visible blocks are not correctly loaded, since the camera pitch and yaw sent to the server are not adjusted accordingly. This fixes that.

### CODE REVIEW

File: src/client.cpp [hunk 1]
```diff
@@ -1307,21 +1311,27 @@ void Client::sendPlayerPos()
 
 	u8 camera_fov    = map.getCameraFov();
 	u8 wanted_range  = map.getControl().wanted_range;
+	float pitch = myplayer->getPitch();
+	float yaw = myplayer->getYaw();
+	if (map.getCameraFrontView()) {
+		yaw += 180 * 100;
```
Comment (lhofhansl) [inline src/client.cpp]: The math here is not strictly necessary; as long as the new and old values differ it's sufficient, but this way it's identical to the transformations in writePlayerPos.

**Comment by lhofhansl** [comment]
An easy way to see the problem is to set the camera into 3rd-front view and then walk backwards.
The server will incorrectly send the block in front of you. The blocks behind you are not retrieved.

**Comment by paramat** [comment]
This seems necessary, 3rd person front view is used a lot in making videos.

**Comment by lhofhansl** [comment]
Let's get it in, then :)

**Comment by paramat** [comment]
:-1: I changed my mind, sorry, as this is adding code and a fifth parameter to camera update for something not necessary.
When someone does a front view shot it's usually just after they have been facing the opposite way loading in a nice background for that shot, so it's not really needed.

**Comment by lhofhansl** [comment]
Well it's still wrong. The server retrieves exactly the opposite blocks instead of the ones actually being looked at.

**Comment by paramat** [comment]
Hm indeed, i edited my last post, i'm not sure now, i guess i'm neutral at the moment.

**Comment by lhofhansl** [comment]
I find the entire way this is handled weird. The camera has a direction, but pitch and yaw are handled separately in game.cpp. So the camera direction and pitch/yaw can be out of sync. Redoing that is a bigger fix.

**Comment by est31** [comment]
I'm not sure, but doesn't the code also have to include usage of different camera positions?

**Comment by lhofhansl** [comment]
@est31 I don't think I understand what you mean.

**Comment by est31** [comment]
@lhofhansl well, in 3rd camera view the position of the camera is not the position of the player anymore, but close to it.

**Comment by paramat** [comment]
We have an API that can set different camera offsets from the player.

**Comment by lhofhansl** [comment]
Hmm... It seems the server needs to know both, the player and the camera position.
This fixes one glaring problem without changing the protocol.

**Comment by est31** [comment]
@lhofhansl I agree it improves the current state so :+1:  to this PR. The camera position th

**Comment by lhofhansl** [comment]
Cool. For the position stuff. The server needs the actual position of the player (to check reach, collisions, etc), and it needs the camera position/direction to retrieved the right blocks. Sending both is a lot more data, though.

**Comment by est31** [comment]
Sorry, its only a concept approval, haven't looked at the code (yet).

**Comment by lhofhansl** [comment]
All good. I am not emotionally attached to getting this in in the current form. :)

**Comment by paramat** [comment]
Oops sorry.

**Comment by lhofhansl** [comment]
What should I do with this one? The more I think about it, the more it seems bad that the server loads those blocks that are precisely on the opposite side of what the players sees.

**Comment by paramat** [comment]
It looks like est31 will approve once he reviews. I think it's reasonable but don't feel strongly about it. So there's a good chance of this being approved for merge.

**Comment by Fixer-007** [comment]
I've noticed that too while doing the selfie in minetest, I support this.

**Comment by paramat** [comment]
:+1:

File: src/client.cpp (hunk 1 above)
Comment (SmallJoker) [inline src/client.cpp]: Pitch changes, thus it's not neccessary to change yaw too.

Comment (paramat) [inline src/client.cpp]: How can pitch replace yaw?

Comment (paramat) [inline src/client.cpp]: It seems myplayer->getYaw() returns yaw not encoded with * 100?

Comment (SmallJoker) [inline src/client.cpp]: Should be `180.0` in this place, without multiplier. Discussion reference: http://irc.minetest.net/minetest-dev/2017-03-23#i_4837454

Comment (paramat) [inline src/client.cpp]: Yes, see https://github.com/minetest/minetest/blob/master/src/content_cao.cpp#L486

**Comment by paramat** [comment]
@lhofhansl +1 once the * 100 issue is fixed and retested.

**Comment by lhofhansl** [comment]
Haven't looked at this for a while, I faintly remember that I did find some implication with 3rd person view.

**Comment by lhofhansl** [comment]
I think that's fine, because there _is_ no correct tilt of the head to look at what the player is looking at.
But somebody should try it out: Install the playeranim mod and then try the 3rd person view with and without this patch.

Without the patch your player would always look back in exactly your direction. With the patch the player looks up when you look up and down when you look down. It's hard to describe.

**Comment by lhofhansl** [comment]
Fixed the *100 issue. Excellent catch, btw. And rebased.
If somebody could check out the playeranim issue and confirm it's cool then we should merge this.

**Comment by paramat** [comment]
:+1:

**Comment by paramat** [comment]
Would the playeranim mod be fixable though? If so that's a mod issue that shouldn't block this.

**Comment by lhofhansl** [comment]
The other weirdness is that if you look at another player the direction you see that player looking (with or without playeranim) is the direction on which the server "thinks" it is looking.

So if you go to 3rd person view, other players will see you as looking into the other direction. In a sense it is correct, since it _is_ the direction you're looking at right now, but the player is also turned.

**Comment by lhofhansl** [comment]
Can somebody try this out? The real problem is that the server cannot distinguish the orientation and position of the camera and player. To do it right we'd need to send both.

With this patch the player will appear to other players as if oriented the way the camera looks, which is not wrong, but also not exactly right.

The question is, is it worse that the server will send a complete worthless set of blocks and not send the blocks the player actually sees.

**Comment by paramat** [comment]
Yes i am unsure now if the small advantage is worth the cost of the minor bugs.

> the server cannot distinguish the orientation and position of the camera and player. To do it right we'd need to send both.

I would prefer to not do this.

**Comment by lhofhansl** [comment]
Perhaps instead of inverting the camera at the client, we send a bit along that the camera is inverted. Then the server can do the right thing. It's a protocol change, though, but it will just need to send a bit (i.e. an extra byte on the wire).

File: src/client.cpp [hunk 2]
```diff
@@ -1027,6 +1027,10 @@ void writePlayerPos(LocalPlayer *myplayer, ClientMap *clientMap, NetworkPacket *
 	u8 wanted_range  = MYMIN(255,
 			std::ceil(clientMap->getControl().wanted_range / MAP_BLOCKSIZE));
 
+	if (clientMap->getCameraFrontView()) {
+		yaw += 180 * 100;
```
Comment (nerzhul) [inline src/client.cpp]: can you add a define for 180 to define what is this and use it in the multiple places ?

**Comment by SmallJoker** [comment]
Tested it and well.. it loads the blocks correctly. I'm not sure if it's a good idea to send another player rotation than it actually is.

File: src/client.cpp (hunk 2 above)
Comment (paramat) [inline src/client.cpp]: Why? It's not a constant that may need changing? -1

**Comment by paramat** [comment]
> Perhaps instead of inverting the camera at the client, we send a bit along that the camera is inverted. Then the server can do the right thing.

I'm not keen on this.
People tend to use front third view when all around them is loaded, and walking backwards long distance in this view is very rare, i don't think it's worth doing since it causes new types of weirdness.

**Comment by lhofhansl** [comment]
Well... If I do it as described it would _not_ include any weirdness.

**Comment by paramat** [comment]
Sorry you're right, i was confused.

What i mean is that this seems too minor and rare an issue to send extra information sent (not much i know) and a do a protocol bump. As you wrote in your 1st post: "not sure it's even worth the effort".

**Comment by lhofhansl** [comment]
I can check the size of the packet and optionally use an extra byte or not, we do this in other cases too.
While the effect at this point is just mildly annoying, it's still plain and simply wrong behavior. The server should know where the player is and in which direction the camera is pointing.

**Comment by paramat** [comment]
Low priority, not for 0.4.16 i feel.

**Comment by paramat** [comment]
Well, i'm fairly neutral. protocol bump is no issue as we are breaking stuff for 0.5.0 currently.

By the way, are you still happy to be a new core dev? We've just lost another one and the lack of dev time is quite severe recently. There are no obligations for a certain rate of work, you can just work on stuff when you feel like it, i know you're busy. What we need is the ability to review C++ code, and someone to test PRs. Your contributions have been excellent and we're impressed by those.

**Comment by lhofhansl** [comment]
> By the way, are you still happy to be a new core dev?

I'd still be honored. Work keeps me busy and I'm going through some private changes. With those caveats, sure thing!

**Comment by lhofhansl** [comment]
As for this issue. Reflecting on this. Simply inverting the camera is a hack and leads to other annoying issues.

I think we should send another bit to indicate that the camera is inverted. That way the player appears correct, and the service can avoid retrieving a useless set of blocks but instead actually those the client is looking at :)

**Comment by paramat** [comment]
Sorry for my earlier resistence, i'm neutral. Needs input from others.

**Comment by celeron55** [comment]
@lhofhansl Can you join Freenode sometime and PM me so that we can handle the core dev details?

**Comment by lhofhansl** [comment]
Rebased. Will add changes soon to send an extra bit along.

**Comment by paramat** [comment]
@celeron55 i asked all active devs, all are +1 or neutral about lhofhansl becomng a core dev.

**Comment by lhofhansl** [comment]
OK... Updated with sending a bit to the server. Still a simple patch, but the size of it is annoying - many places to track this one bit.

**Comment by lhofhansl** [comment]
Comments on the change? (in checks fail on code that this PR is not touching...?)

I also hung out on freenode for a bit, but @celeron55 wasn't there :)

**Comment by paramat** [comment]
Bump for reviews.

File: src/clientmap.h [hunk 3]
```diff
@@ -115,6 +116,7 @@ class ClientMap : public Map, public scene::ISceneNode
 
 	const MapDrawControl & getControl() const { return m_control; }
 	f32 getCameraFov() const { return m_camera_fov; }
+	bool getCameraInverted() const { return m_camera_inverted; }
```
Comment (rubenwardy) [inline src/clientmap.h]: `getIsCameraInverted`

File: src/clientmap.h [hunk 4]
```diff
@@ -127,6 +129,7 @@ class ClientMap : public Map, public scene::ISceneNode
 	v3f m_camera_direction = v3f(0,0,1);
 	f32 m_camera_fov = M_PI;
 	v3s16 m_camera_offset;
+	bool m_camera_inverted;
```
Comment (rubenwardy) [inline src/clientmap.h]: `m_is_camera_invented`

**Comment by paramat** [comment]
@lhofhansl see above, you'll have a +1 with those minor changes.

**Comment by paramat** [comment]
@lhofhansl poke. See above.

File: src/clientmap.h (hunk 4 above)
Comment (paramat) [reply src/clientmap.h]: LOL.

**Comment by paramat** [comment]
Closing due to 1-month rule, we'll re-open on progress.

**Comment by paramat** [comment]
@lhofhansl see above, you'll have a +1 and can merge with those minor changes.

**Pull request closed without merging**